pkg/pool-weighted/contracts/smart/ManagedPool.sol

```diff
@@ -405,34 +386,151 @@ contract ManagedPool is BaseWeightedPool, ReentrancyGuard {
+            preSwapBalances[1],
+            normalizedWeights[1],
+            amountInMinusSwapFees
+        );
+
+        uint256[] memory postSwapBalances = new uint256[](2);
+        postSwapBalances[0] = preSwapBalances[0].add(request.amount);
+        postSwapBalances[1] = preSwapBalances[1].sub(amountOut);
+
+        _payProtocolAndManagementFees(normalizedWeights, preSwapBalances, postSwapBalances);
 
-        return super._onSwapGivenOut(swapRequest, currentBalanceTokenIn, currentBalanceTokenOut);
+        // amountOut tokens are exiting the Pool, so we round down.
+        return _downscaleDown(amountOut, scalingFactorTokenOut);
     }
 
-    /**
-     * @dev Used to adjust balances by subtracting all collected fees from them, as if they had been withdrawn from the
-     * Vault.
-     */
-    function _subtractCollectedFees(uint256[] memory balances) private view {
-        for (uint256 i = 0; i < _getTotalTokens(); ++i) {
-            // We can use unchecked getters as we know the map has the same size (and order!) as the Pool's tokens.
-            balances[i] = balances[i].sub(_tokenCollectedManagementFees.unchecked_valueAt(i));
+    function _processSwapGivenOut(
+        SwapRequest memory request,
+        uint256 scalingFactorTokenIn,
+        uint256 scalingFactorTokenOut,
+        uint256[] memory preSwapBalances,
+        uint256[] memory normalizedWeights
+    ) private returns (uint256) {
+        request.amount = _upscale(request.amount, scalingFactorTokenOut);
+
+        uint256 amountIn = WeightedMath._calcInGivenOut(
+            preSwapBalances[0],
+            normalizedWeights[0],
+            preSwapBalances[1],
+            normalizedWeights[1],
+            request.amount
+        );
+
+        // amountIn tokens are entering the Pool, so we round up.
+        amountIn = _downscaleUp(amountIn, scalingFactorTokenIn);
+
+        // Fees are added after scaling happens, to reduce the complexity of the rounding direction analysis.
+        uint256 amountInPlusSwapFees = _addSwapFeeAmount(amountIn);
+
+        uint256[] memory postSwapBalances = new uint256[](2);
+        postSwapBalances[0] = _upscale(amountInPlusSwapFees, scalingFactorTokenIn);
+        postSwapBalances[1] = preSwapBalances[1].sub(request.amount);
+
+        _payProtocolAndManagementFees(normalizedWeights, preSwapBalances, postSwapBalances);
+
+        return amountInPlusSwapFees;
+    }
+
+    function _payProtocolAndManagementFees(
+        uint256[] memory normalizedWeights,
+        uint256[] memory preSwapBalances,
+        uint256[] memory postSwapBalances
+    ) private {
+        // Calculate total BPT for the protocol and management fee
+        // The management fee percentage applies to the remainder,
+        // after the protocol fee has been collected.
+        // So totalFee = protocolFee + (1 - protocolFee) * managementFee
+        uint256 protocolFeePercentage = paysProtocolFees() ? _cachedProtocolSwapFeePercentage : 0;
```

Comment:
Well, it would be more gas per swap (external call vs storage read) - but then we wouldn't need the external call to update it. Protocol fees are maxed now, but if they were ever lowered, and nobody remembered to call this, it would overcollect fees for a while. I'll measure the gas and see how bad it is.

Reply:
It costs 3,269 more in gas